src/server/conn_context.h

```diff
@@ -17,20 +18,24 @@ class EngineShardSet;
 class ConnectionContext;
 class ChannelStore;
 
+// Stores command id and arguments for delayed invocation.
+// Used for storing MULTI/EXEC commands.
 struct StoredCmd {
-  const CommandId* descr;
+  StoredCmd(const CommandId* cid, CmdArgList args);
 
- private:
-  std::unique_ptr<char[]> backing_storage_;
-  CmdArgVec arg_vec_;
-  CmdArgList arg_list_;
+  size_t NumArgs() const;
 
- public:
-  StoredCmd(const CommandId* d, CmdArgList args);
+  // Fill the arg list with stored arguments, it should be at least of size NumArgs() + 1.
+  // cmd_scratch is used for storing the command name in a mutable slice (can be null to skip).
+  // Between filling and invocation, the stored cmd and the scratch should NOT be moved.
+  void Fill(CmdArgList args, MutableSlice cmd_scratch);
 
-  CmdArgList ArgList() const {
-    return arg_list_;
-  }
+  const CommandId* Cid() const;
+
+ private:
+  const CommandId* cid_;                 // underlying command
+  std::string buffer_;                   // underlying buffer
+  absl::FixedArray<uint32_t, 4> sizes_;  // sizes of arg parts
```

Comment:
_Ideally_, all this should be just a variant<char[64], unique_ptr<char[]>> with \0 for divisions, but it looks a bit over-engineered. Wdyt? 

We waste space on the pointer (8 bytes - longer that many commands) and the sizes (most of them are small, wasted 3 bytes), so the current SOO is ~20 chars + 4 args with a struct size of 64 😮‍💨 

Again, its not noticeable with multi/exec, but if I do

```lua
for i=1,100 do
  redis.acall('LPUSH', 'l', 1)
end
```

and bombard with with memtier, then mimialloc new/free takes up to 10-15%.

Reply:
Yes, I think we can stop here :)